libs/MeshKernel/src/CurvilinearGrid/CurvilinearGrid.cpp

```diff
@@ -1304,9 +1310,91 @@ std::vector<CurvilinearGrid::CurvilinearFaceNodeIndices> CurvilinearGrid::Comput
 }
 
+std::set<CurvilinearGrid::CurvilinearEdge> CurvilinearGrid::ComputeBoundaryEdges(const CurvilinearGridNodeIndices& lowerLeft, const CurvilinearGridNodeIndices& upperRight) const
+{
+    std::vector<Point> result;
+    const auto facesIndices = ComputeFaceIndices(lowerLeft, upperRight);
+    std::set<CurvilinearEdge> boundaryEdges;
+
+    for (const auto& faceIndices : facesIndices)
+    {
+        if (!std::ranges::all_of(faceIndices, [this](const CurvilinearGridNodeIndices& curviNode)
+                                 { return GetNode(curviNode.m_n, curviNode.m_m).IsValid(); }))
+        {
+            continue;
+        }
+
+        const auto numFaceNodes = static_cast<UInt>(faceIndices.size());
+        for (UInt i = 0u; i < numFaceNodes; ++i)
+        {
+            const auto firstCurvilinearNodeIndex = faceIndices[i];
+            const auto nextIndex = NextCircularForwardIndex(i, numFaceNodes);
+            const auto secondCurvilinearNodeIndex = faceIndices[nextIndex];
+
+            const auto edge = std::make_pair(std::min(firstCurvilinearNodeIndex, secondCurvilinearNodeIndex),
+                                             std::max(firstCurvilinearNodeIndex, secondCurvilinearNodeIndex));
+
+            const auto it = boundaryEdges.find(edge);
+            if (it != boundaryEdges.end())
+            {
+                boundaryEdges.erase(it);
+            }
+            else
+            {
+                boundaryEdges.insert(edge);
+            }
+        }
+    }
+    return boundaryEdges;
+}
+
+std::vector<meshkernel::Point> CurvilinearGrid::ComputeBoundaryToPolygon(const CurvilinearGridNodeIndices& lowerLeft, const CurvilinearGridNodeIndices& upperRight) const
+{
+    std::vector<Point> result;
+
+    auto boundaryEdges = ComputeBoundaryEdges(lowerLeft, upperRight);
+    if (boundaryEdges.empty())
+    {
+        return result;
+    }
+
+    auto currentEdge = boundaryEdges.begin();
+    auto startNode = currentEdge->first;
+    auto sharedNode = currentEdge->second;
+
+    while (!boundaryEdges.empty())
+    {
+        result.push_back(GetNode(startNode.m_n, startNode.m_m));
+        while (sharedNode != startNode)
+        {
+            result.push_back(GetNode(sharedNode.m_n, sharedNode.m_m));
+
+            boundaryEdges.erase(currentEdge);
+            currentEdge = std::ranges::find_if(boundaryEdges, [&](const CurvilinearEdge& edge)
+                                               { return edge.first == sharedNode ||
+                                                        edge.second == sharedNode; });
+
+            sharedNode = currentEdge->first == sharedNode ? currentEdge->second : currentEdge->first;
+        }
+        result.push_back(GetNode(startNode.m_n, startNode.m_m));
+        boundaryEdges.erase(currentEdge);
+        if (boundaryEdges.empty())
+        {
+            return result;
+        }
+
+        result.emplace_back(constants::missing::doubleValue, constants::missing::doubleValue);
```

Comment:
add a comment: next disjunct loop starts again with smallest node index of remaining edges